### PR TITLE
Nested decks in pull request

### PR DESCRIPTION
**Pull request by Arthur-Milchior**
## Purpose / Description

Adding nested decks to test. Seems especially useful since it seems they failed and required some change

## Fixes

They seemed to be requested in 
https://github.com/ankidroid/Anki-Android/pull/6526#pullrequestreview-434534141

## Approach
First: adding a nested deck. Running test. Seeing test fails, and correcting them in a way which makes sens

## How Has This Been Tested?

Running test.

## Learning (optional, can help others)

I don't need to understand all matrix and all to run tests

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

File: AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.java [hunk 1]
```diff
@@ -118,9 +125,21 @@ public void setUp() throws Exception {
...
         mNumDecksBeforeTest = col.getDecks().count();
         for(int i = 0; i < TEST_DECKS.length; i++) {
-            long did = col.getDecks().id(TEST_DECKS[i]);
-            mTestDeckIds[i] = did;
-            mCreatedNotes.add(setupNewNote(col, mModelId, did, mDummyFields, TEST_TAG));
+            String fullName = TEST_DECKS[i];
+            String[] path = col.getDecks().path(fullName);
+            String partialName = "";
+            // Dealing with each parents of fullName
+            for (int j = 0; j < path.length; j++) {
```
Comment (david-allison-1) [inline AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.java:132]: Why is this necessary? Shouldn't `id()` call `_ensureParents`? Is this because decks are created, but not added to the list?

Comment (Arthur-Milchior) [reply AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.java:132]: id ensure parents, indeed. But if you want mTestDeckIds to contains all ids, then you need to put ids of parents too, even if they are not explicitly in the list of parents.
Furthermore, during tear-down, you want to remove all added deck; and you have no method "_ensureRemoveParent"

Comment (david-allison-1) [reply AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.java:132]: That makes sense. Could you add a quick comment

**Comment by Arthur-Milchior** [comment]
Pushed with a new comment, as requested by @david-allison-1 . No other change

**Comment by Arthur-Milchior** [comment]
testQueryAllDecks() seems to be exactly what you're looking for. Except that it didn't used to test for nested deck since there were none in the test areay

**Comment by david-allison-1** [comment]
> testQueryAllDecks() seems to be exactly what you're looking for. Except that it didn't used to test for nested deck since there were none in the test areay

I missed that. Thanks!

As an interesting side-note. There are upsides and downsides to checking a system against itself, instead of against predefined test data. 

For example, changing `deck.getString("name")` to return "Z" for a certain input likely wouldn't be caught with that test, as both the API and the `Decks` class would be affected.

That being said, I'm confident that the area's now tested against nested decks. Thanks!

**Comment by Arthur-Milchior** [comment]
I mean, this is testing "query deck", quite literally. It is not a coverage of the whole Deck manager.
This return the list of decks. I can't test it against a fixed value, because it also returns the decks currently in the collection on the device. We could potentially do it the day were tests are done in a new collection instead of reusing the user's collection. Alas, I fail to figure out how to do it without changing non-test code.

**Comment by mikehardy** [comment]
Use the unit testing vs the androidTest instrumented testing.

Make a class that extends RobolectricTest (if one like 'DecksTest' doesn't exist already)

in your test case set up your data hierarchy from scratch (it will be in a throwaway in-memory database, starting empty)

total control of the data, and you can assert that data in == data out and garbage in == exceptions :)
